### PR TITLE
Add OpenGL ES support

### PR DESCRIPTION
**Pull request by caramelli**
This change adds a basic OpenGL ES support.

Nicolas Caramelli

### CODE REVIEW

File: base/OpenGLVisitor.cpp
Comment (rzr) [inline base/OpenGLVisitor.cpp:123]: I had in mind to duplicate OpenGLVisitor in OpenGLESVisitor it could be easier to change renderer at runtime eventually ?

File: base/config-rzr.h
Comment (rzr) [inline base/config-rzr.h:199]: did you try to build it under msvc ?
if not please preserve this change , 
and please commit a cleanup patch separately

File: configure.ac
Comment (rzr) [inline configure.ac:136]: this is harmless to keep no?

**Review by rzr — COMMENTED** [review]
Excellent ! I did not test it yet but it looks good,
I suggested to separate cleanup changes if you can
I'll merge if no objections.

File: base/OpenGLVisitor.cpp
Comment (caramelli) [reply base/OpenGLVisitor.cpp:123]: Yes, why not, the code will certainly be better (I'm not a C++ expert!). But to change renderers at runtime, I think OpenGL libraries need to be loaded dynamically.

File: configure.ac
Comment (caramelli) [reply configure.ac:136]: With the proposed change, LIBS="$GL_LIBS $LIBS" is required.

File: base/config-rzr.h
Comment (caramelli) [reply base/config-rzr.h]: I haven't tried with Windows and Visual so I removed cleanup changes as suggested.

**Comment by caramelli** [comment]
I removed cleanup changes as suggested. I'd rather let you make those cleanup changes.
Before merging, can you first test it on your side? I'm curious about the performance you get.

**Comment by rzr** [comment]
Thx I tried that patch on my desktop with nvidia card but only fonts are displaying, maybe my base is dirty, I'll try again on Pi.

Other suggestion may I suggest to try to relocate for of your opengl #ifdef to OpenGLRenderer class (or a newer one like OpenGLESRenderer to preserve the desktop one).

You can start by create a Renderer::setFrustum() that will remove glFrustum from TextureUtils 

what do you think ?

**Comment by caramelli** [comment]
The specificity related to OpenGL ES 1.1 CM seems really minor to justify abstracting this. But depending on what you plan to refactor in the future, feel free to do what's best for you.

If I'm not wrong, the indentation is consistent with the code already there. But I agree that the indentation is not always perfect (this is prior to my PR): yes, it would be nice to have a separate commit to fix it.

File: base/Config.h
Comment (rzr) [inline base/Config.h:28]: ok here but I think this should be done in SDL2, i'll try to open a PR there if you don't

```
/usr/include/SDL/SDL_opengl.h:#include <OpenGL/gl.h>    /* Header File For The OpenGL Library */
/usr/include/SDL/SDL_opengl.h:#include <OpenGL/glu.h>   /* Header File For The GLU Library */
/usr/include/SDL/SDL_opengl.h:#include <gl.h>           /* Header File For The OpenGL Library */
/usr/include/SDL/SDL_opengl.h:#include <glu.h>  /* Header File For The GLU Library */
/usr/include/SDL/SDL_opengl.h:#include <GL/gl.h>        /* Header File For The OpenGL Library */
/usr/include/SDL/SDL_opengl.h:#include <GL/glu.h>       /* Header File For The GLU Library */
```

File: base/OpenGLVisitor.cpp
Comment (rzr) [inline base/OpenGLVisitor.cpp:22]: please keep this I can be cleaned up in other change because I've seen ifdef EPENGL_LIGHTS elsewhere

File: base/OpenGLVisitor.cpp
Comment (rzr) [inline base/OpenGLVisitor.cpp:18]: not related to gles port please keep it

File: base/OpenGLVisitor.cpp
Comment (rzr) [inline base/OpenGLVisitor.cpp:127]: can this dynamic allocation relocated elswhere (use C++ new is prefered)

File: base/OpenGLVisitor.cpp
Comment (rzr) [inline base/OpenGLVisitor.cpp:179]: An other option would be to use display lists ?

File: base/OpenGLVisitor.cpp
Comment (rzr) [inline base/OpenGLVisitor.cpp:575]: delete []

File: base/OpenGLVisitor.cpp
Comment (rzr) [inline base/OpenGLVisitor.cpp:559]: float verts[] = new float[ ... ]

**Review by rzr — APPROVED** [review]
Unless you want to take my remarks into account I'll merge it and eventually refactor it a bit 

For the record I have logged some results at:

https://github.com/adoptware/pinball/issues/22


something else is missing:

```./base/TextureUtil.cpp\0183:  SDL_GL_SetAttribute( SDL_GL_DOUBLEBUFFER, 1 );
./base/TextureUtil.cpp\0222:  SDL_GL_GetAttribute( SDL_GL_DOUBLEBUFFER, &value );
./base/TextureUtil.cpp\0223:  cerr << "SDL_GL_DOUBLEBUFFER: " << value << endl << endl;
```

This is only for GL/gl.h users

     /usr/include/GL/gl.h:#define GL_DOUBLEBUFFER                            0x0C32

**Comment by caramelli** [comment]
For SDL_GL_GetAttribute( SDL_GL_DOUBLEBUFFER, &value ), this is an issue in SDL2 (this does not happen when building SDL2 without OpenGL desktop support). Maybe I'll send a fix to the SDL2 project for this. I'm not sure that it's really useful to print this information, but I'm leaving it for the OpenGL desktop if you'd prefer to keep it.

Nicolas Caramelli

File: base/Config.h
Comment (caramelli) [reply base/Config.h]: You're right, it's more consistent. Thanks.

File: base/OpenGLVisitor.cpp
Comment (caramelli) [reply base/OpenGLVisitor.cpp]: There is OPENGL_LIGHTS in AmbientLightVisitor.cpp and TextureUtil.cpp, I see EM_OPENGL_LIGHTS only in OpenGLVisitor.cpp and it doesn't seem to be used.
glEnableClientState() is used in the OpenGL ES 1.1 CM port but inside the EM_VERTEXARRAY flag I don't really understand the purpose.
As you wish, we can keep it for now.

File: base/OpenGLVisitor.cpp
Comment (caramelli) [reply base/OpenGLVisitor.cpp]: As you wish, we can keep it for now.

File: base/OpenGLVisitor.cpp
Comment (caramelli) [reply base/OpenGLVisitor.cpp]: Don't hesitate to update it if you are planning a refactoring.

File: base/OpenGLVisitor.cpp
Comment (caramelli) [reply base/OpenGLVisitor.cpp]: OK

File: base/OpenGLVisitor.cpp
Comment (caramelli) [reply base/OpenGLVisitor.cpp]: It's done.

**Review by rzr — APPROVED** [review]
I'll open a new PR with refactoring 
Thx

**Pull request merged**
